### DISCUSSION TITLE
Integration of DPC++ and openCL implementation

### DISCUSSION THREAD

**Discussion by tneval** [intel/llvm#14864] [Q&A] [answered]
Hi,
I've been wondering how exactly DPC++ ends up calling the OpenCL API.
For example, the clCreateProgramWithIL() should be provided by the OpenCL implementation and it is accessible through libOpenCL.so. 
I have found that DPC++ utilizes Plugin Interface, but how it works, is not clear to me.
By looking at the DPC++ sources, there seems to be /plugins/opencl/pi_opencl.cpp, which contains:
![Screenshot from 2024-07-31 17-26-32](https://github.com/user-attachments/assets/a5f607de-787f-4878-b351-b25fe53f378f)

Leading to:

![Screenshot from 2024-07-31 17-26-17](https://github.com/user-attachments/assets/5b0c428a-ff6d-4c52-891b-366236cabdf5)

And finally:
![Screenshot from 2024-07-31 17-29-35](https://github.com/user-attachments/assets/6f19809d-15b4-4ce1-8403-f4ef05878d30)

But I don't understand how this leads to OpenCL API call. Could someone clarify this?

Thanks,

Tapio

**Comment by AlexeySachkov** [comment, marked as answer]
Hi @tneval, you caught our codebase in transition period and that's why it is even more confusing than usual :)

Yes, we used to have Plugin Interface, which would abstract low-level runtimes from us allowing to dynamically select which low-level runtime we would like to use. You can find documentation for PI layer [here](https://github.com/intel/llvm/blob/c0c4d02abc7f6cc167628cf9b411a4354c8f941e/sycl/doc/design/PluginInterface.md). Note that this links points to an older commits, because that file does not exists anymore, details below.

PI essentially evolved into its own standalone project called [Unified Runtime](https://github.com/oneapi-src/unified-runtime). What you see is that all of our PI plugins are being redirected into Unified Runtime, where Unified Runtime will then redirect that call to a certain adapter (like OpenCL one) and do `cl*` call there.

If you take a look at the most recent codebase (the final switchover happened last Friday), then you won't see `PI` layer anymore: SYCL RT now calls directly into Unified Runtime which is then dispatches the call into one of adapters, so hopefully it is now less confusing.
Documentation for UR can be found [here](https://oneapi-src.github.io/unified-runtime/core/INTRO.html)

**Comment by tneval** [comment]
Thanks, very helpful!
